### ISSUE TITLE
Question: Does Parallel-consumer have state that we can read from?

### ISSUE THREAD

**Issue by Ehud-Lev-Forter** [confluentinc/parallel-consumer#484]
Hi, We have noticed that during high load, one of the consumers can get stuck, meaning it will report the same lag, but will not process anything, we were not able to reproduce it locally, but we are afraid that it will happen again. 

In our kafka-stream process we supervise [state](https://github.com/apache/kafka/blob/c8a228d122b32d9df9f56b43fe55a27a07ee6a12/streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java#L360) to monitor the process, if we see an error or problem, we restart the process. Does ParallelConsumer have something similar?

**Comment by astubbs** [comment]
Hi, we’re developing the metrics and monitoring systems now:
- #464 
- #34 
- #27 

What were the logs saying when this happened?
Are you checking the record context for failure attempts in your user function?

Lag is a bit more complicated with PC. For example, if there’s a single poison pill that can’t process and your user function doesn’t give up retrying it, lag could show 1,000 but there may only be a single record that is “stuck”.

The new monitoring and metrics functions will allow you to see all this info. 

Cc @nachomdo

**Comment by Ehud-Lev-Forter** [comment]
Hi, thank you for your response,
We did get this issue again, when it happened we saw many warning logs that looked like the following:
`Warning: 1 records in the queue have been waiting longer than 10s for following topics [my-topic-2].` 
Important to mention that other processes also got those warnings but much less. 
We are using version `0.5.2.4` with the "`reactor consumer`" and till now I didn't notice the "`record context`", we used the `context.streamConsumerRecords() `instead, I will check it out and see if I can catch errors through that. 

As for now, the only way for us to resolve is to track the lag (using external metrics) and restart the process.

**Comment by astubbs** [comment]
Are you logging any processing failures in your function?
Do you see any ERROR level events in the log?
If there really is a "stuck" record, it should still get given to the user function to attempt.

**Comment by Ehud-Lev-Forter** [comment]
Yes, I am printing everything I can (error-info), I also trying to catch all exceptions and send records that has failed into different DLQs. I am printing the `record.numberOfFailedAttempt` but I did not get any of those yet. I had the issue again, and I restarted the process again, and I got the following warning after the restart:
`Truncating state - removing records lower than 54845363. Offsets have been removed from the partition by the broker or committed offset has been raised. Bootstrap polled 54845363 but expected 1 from loaded commit data. Could be caused by record retention or compaction and offset reset policy LATEST.","logger_name":"io.confluent.parallelconsumer.state.PartitionState","thread_name":"pc-control","level":"WARN",`

I am now trying to tackle the issue by adding some kind of supervisor that detects that this consumer is not really doing anything and kill it.

**Comment by astubbs** [comment]
What is the head offset of that partition?
What is your auto offset reset policy set to?

**Comment by Ehud-Lev-Forter** [comment]
I don't have the historical information, I can say that now ( 1.5 hours after the restart) it is 
CURRENT-OFFSET  LOG-END-OFFSET  LAG
54203593                 54203720               127
Must say that this does not make sense to me since it seems that it complained on future offset.

As for the auto offset reset policy `earliest`

**Comment by astubbs** [comment]
> Update: It might be that this log was talking on another partition, this process consumed from another partition as well: 55016222 55016324

ah - does the log message not include the partition it's talking about?

@Ehud-Lev-Forter , if you get a chance, take a look at this PR and see if there's any other metrics you'd like?

#485

**Comment by astubbs** [comment]
cc @nachomdo

**Comment by Ehud-Lev-Forter** [comment]
Looks like we can benefit a lot from this PR as is.
As for other metrics, I am not sure if it makes sense, but I can think those:
1. Throttling or rate limit, I saw that PC push back some work when things get slow.
2. Maybe slowWork.size gauge.

As for the warn log, yes it is missing the partition log

**Comment by astubbs** [comment]
1 - yup, that status will be in there
2 - oh - that's a good idea!

**Comment by astubbs** [comment]
Tracked in the maintained fork as astubbs/parallel-consumer#157, which carries a summary of this
issue and its current status there.

The **Confluent-hosted version** of this project is no longer maintained - its README says so.
**Parallel Consumer itself continues** in my fork: https://github.com/astubbs/parallel-consumer

Nothing in this repository is closed or changed by this comment.
